jni/src/faiss_index_service.cpp

```diff
@@ -250,5 +250,120 @@ void BinaryIndexService::writeIndex(
+    if(threadCount != 0) {
+        omp_set_num_threads(threadCount);
+    }
+
+    // Add extra parameters that cant be configured with the index factory
+    SetExtraParameters<faiss::Index, faiss::IndexIVF, faiss::IndexHNSW>(jniUtil, env, parameters, index.get());
+
+    // Check that the index does not need to be trained
+    if(!index->is_trained) {
+        throw std::runtime_error("Index is not trained");
+    }
+
+    std::unique_ptr<faiss::IndexIDMap> idMap (faissMethods->indexIdMap(index.get()));
+    //Makes sure the index is deleted when the destructor is called, this cannot be passed in the constructor
+    idMap->own_fields = true;
+
+    allocIndex(dynamic_cast<faiss::Index *>(idMap->index), dim, numVectors);
+
+    //Release the ownership so as to make sure not delete the underlying index that is created. The index is needed later
+    //in insert and write operations
+    index.release();
+    return reinterpret_cast<jlong>(idMap.release());
+}
+
+void ByteIndexService::insertToIndex(
+        int dim,
+        int numIds,
+        int threadCount,
+        int64_t vectorsAddress,
+        std::vector<int64_t> & ids,
+        jlong idMapAddress
+    ) {
+    // Read vectors from memory address
+    auto *inputVectors = reinterpret_cast<std::vector<int8_t>*>(vectorsAddress);
+
+    // The number of vectors can be int here because a lucene segment number of total docs never crosses INT_MAX value
+    int numVectors = inputVectors->size() / dim;
+    if(numVectors == 0) {
+        throw std::runtime_error("Number of vectors cannot be 0");
+    }
+
+    if (numIds != numVectors) {
+        throw std::runtime_error("Number of IDs does not match number of vectors");
+    }
+
+    // Set thread count if it is passed in as a parameter. Setting this variable will only impact the current thread
+    if(threadCount != 0) {
+        omp_set_num_threads(threadCount);
+    }
+
+    faiss::IndexIDMap * idMap = reinterpret_cast<faiss::IndexIDMap *> (idMapAddress);
+
+    // Add vectors in batches by casting int8 vectors into float with a batch size of 1000
+    int batchSize = 1000;
+    std::vector <float> inputFloatVectors(batchSize * dim);
+    std::vector <int64_t> floatVectorsIds(batchSize);
+    int id = 0;
+    auto iter = inputVectors->begin();
+
+    for (int id = 0; id < numVectors; id += batchSize) {
+        if (numVectors - id < batchSize) {
+            batchSize = numVectors - id;
+            inputFloatVectors.resize(batchSize * dim);
+            floatVectorsIds.resize(batchSize);
+        }
+
+        for (int i = 0; i < batchSize; ++i) {
+            floatVectorsIds[i] = ids[id + i];
+            for (int j = 0; j < dim; ++j, ++iter) {
+                inputFloatVectors[i * dim + j] = static_cast<float>(*iter);
+            }
+        }
+        idMap->add_with_ids(batchSize, inputFloatVectors.data(), floatVectorsIds.data());
```

Comment:
You can use float[] and int64_t[] here. and then you would never need to do resize which you are doing in line number 340 and 341. 

As resize has Linear complexity and can trigger a reallocation of memory, which will further spike the memory consumption. Ref: https://cplusplus.com/reference/vector/vector/resize/

Also if you see you don't even need resize, as you updating the vectors at same reference and updating the batchSize attribute. 
In that case even when `inputFloatVectors` is lets say has size 100, but batchSize is 10, then only 10 elements will be picked from inputFloatVectors. You can check in Faiss. Last I checked I remember something like this.

```
idMap->add_with_ids(batchSize, inputFloatVectors.data(), floatVectorsIds.data());
```

Reply:
Tested with arrays, which was running into segmentation fault with benchmarking test. We can avoid resize with std::vector which works based on batchsize and I have tested it
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f20a93e33a0, pid=80, tid=295
#
# JRE version: OpenJDK Runtime Environment Temurin-21.0.1+12 (21.0.1+12) (build 21.0.1+12-LTS)
# Java VM: OpenJDK 64-Bit Server VM Temurin-21.0.1+12 (21.0.1+12-LTS, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# C  [libopensearchknn_faiss_avx2.so+0x1de3a0]  knn_jni::faiss_wrapper::ByteIndexService::insertToIndex(int, int, int, long, std::vector<long, std::allocator<long> >&, long)+0x210
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/lib/systemd/systemd-coredump %P %u %g %s %t %c %h" (or dumping to /home/ci-runner/opensearch/core.80)
#
# An error report file with more information is saved as:
# logs/hs_err_pid80.log
[23.597s][warning][os] Loading hsdis library failed
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

Reply:
Using malloc instead of `arr[batchSize*dim]` may cause issue as malloc will create array on heap but `arr[batchSize*dim]` create on stack. Please use that.

Reply:
Update: Arrays also will work using malloc
https://github.com/opensearch-project/k-NN/pull/1823#discussion_r1728537543